### ISSUE TITLE
FR: Consider supporting shallow clones of LLVM and Clang

### ISSUE THREAD

**Issue by creachadair** [kythe/kythe#3307]
The LLVM and Clang repositories are very large, both in size and in number of objects. The `tools/modules/update.sh` script currently clones and checks out out the whole repository for each. Both projects have a long history and a lot of code, and everyone who wants to build/develop against Kythe winds up having to download that complete history.

The obvious solution would be to use shallow clones (e.g., `git clone --depth=1 …`), but that does not quite work because we pin (via `versions.sh`) to specific commit hashes. In _principle_ one could fetch just a single commit, but the Git hosting sites (including GitHub) do not permit you to request an arbitrary hash, only "known" commits. As far as I can tell, "known" means commits which are tagged, or are branch heads. Unfortunately, it looks like LLVM doesn't use tags or branches.

We could still make this better for our contributors, though, if  if we are willing to maintain a fork of the mirror and attach tags to our fork. Then we could get a shallow clone of the fork by doing something like:

    git clone --depth 1 --branch v1.2.3 --single-branch $FORKREPO $TARGET
    (cd $TARGET && git checkout -b v1.2.3)

I did this as an experiment, and compare:

```
# github.com/llvm-mirror/llvm
$ du -sh third_party/llvm/llvm
2.3G	third_party/llvm/llvm

# github.com/creachadair/llvm
$ du -sh test/llvm
426M	test/llvm
```

It adds a couple steps to updating the commit hash: Rebasing the fork, and adding/updating the tag. But that happens infrequently enough that it might be worthwhile (and I think it could be scripted).

**Comment by creachadair** [comment]
> We don't have to use `git clone` we can just fetch an http archive of the particular revision, like we do for our other dependencies in WORKSPACE, there's no need to fork.

Does that work for arbitrary commits? That would certainly be easier, but I couldn't see a way to get the exact version we want without a tag or a branch.

**Comment by shahms** [comment]
$ wget https://github.com/llvm-mirror/llvm/archive/e744b686ed7b30e1c6a172ede09f012de33aadac.zip
$ du -hs llvm-e744b686ed7b30e1c6a172ede09f012de33aadac/
386M	llvm-e744b686ed7b30e1c6a172ede09f012de33aadac/

**Comment by shahms** [comment]
there's a missing unzip step there, but yes ;-)

**Comment by creachadair** [comment]
Nevermind, I see @danielmoy-google is already working on it.

**Comment by creachadair** [comment]
Unfortunately, this seems to have introduced some other complications: #3324

**Comment by creachadair** [comment]
Fixed by resolution of #3324.
